src/replication/DealReplicationWorker.ts

```diff
@@ -0,0 +1,355 @@
+      // check if download URL is reachable if necessary
+      if (!model.isOffline && !(await this.checkUrlReachability(downloadUrl))) {
+        throw new Error(`${downloadUrl} is not reachable`);
+      }
+      const priceInAtto = priceInFil.toNumber('am');
+      let propose = `deal --http-url=${downloadUrl}`;
+      if (model.isOffline) {
+        propose = `offline-deal`;
+      }
+      return `${this.boostCMD} ${propose} --provider=${miner}  --commp=${carFile.pieceCid} --car-size=${carFile.carSize} ` +
+        `--piece-size=${carFile.pieceSize} --payload-cid=${carFile.dataCid} --storage-price-per-epoch=${priceInAtto} ` +
+        `--verified=${model.isVerfied} --wallet=${model.client} --duration=${model.duration}`;
+    }
+  }
+
+  /**
+   * Main function of deal making
+   * @param replicationRequest
+   */
+  private async replicate(replicationRequest: ReplicationRequest): Promise<void> {
+    const miners = await this.generateMinersList(replicationRequest.criteria);
+    const boostResultUUIDMatcher = /deal uuid: (\S+)/;
+    let breakOuter = false;
+    for (let i = 0; i < miners.length; i++) {
+      if (breakOuter) {
+        break;
+      }
+      const miner = miners[i];
+      let useLotus = true;
+      try {
+        useLotus = await this.isUsingLotus(miner);
+      } catch (error) {
+        this.logger.error(`SP ${miner} unknown output from libp2p. Give up on this SP.`, error);
+        continue;
+      }
+
+      // Find cars that are finished generation
+      const cars = await Datastore.GenerationRequestModel.find({
+        datasetId: replicationRequest.datasetId,
+        status: 'completed'
+      });
+
+      let dealsMadePerSP = 0;
+      let retryTimeout = config.get<number>('deal_replication_worker.min_retry_wait_ms'); // 30s, 60s, 120s ...
+      for (let j = 0; j < cars.length; j++) {
+        const carFile = cars[j];
+
+        // check if the replication request has been paused
+        const existingRec = await Datastore.ReplicationRequestModel.findById(replicationRequest.id);
+        if (!existingRec) {
+          this.logger.error(`This replication request ${replicationRequest.id} no longer exist.`);
+          breakOuter = true;
+          break;
+        }
+        if (existingRec.status !== 'active') {
+          this.logger.warn(`This replication request ${existingRec.id} has become non-active: ${existingRec.status}.`);
+          breakOuter = true;
+          break;
+        }
+
+        // check if reached max deals needed to be sent
+        if (existingRec.maxNumberOfDeals > 0 && dealsMadePerSP >= existingRec.maxNumberOfDeals) {
+          this.logger.warn(`This SP ${miner} has made max deals planned (${existingRec.maxNumberOfDeals}) by the request ${existingRec.id}.`);
+          break;
+        }
+
+        // check if the car has already dealt or have enough replica
+        const existingDeals = await Datastore.DealStateModel.find({
```

Comment:
i think you don't need to get all existing deals, It's very expensive under a loop. You can use two queries, one is to find deal by provider and pieceCid, and another is to get the count of deals.

Reply:
According to the MongoDB documentation:

[db.collection.count()](https://docs.mongodb.com/manual/reference/method/db.collection.count/)

count() is equivalent to the db.collection.find(query).count() construct.

Reply:
So I think the correct way of handling this is to have compound index.